Automattic/wp-calypso
# DomainPicker: Fix incorrect selected domain on analytics when domain picker closes.

## Changes proposed in this Pull Request

When domain picker is closed, the incorrect `selected_domain` was being tracked on analytics. This was due to the buggy implementation of a duplicated and modified hook `useTrackModal()`. 

This PR proposes the following fix:
* Remove `useTrackModal()` (and along with it, `useOnMount` and `useOnUnmount`) hook from domain picker package. 
* Remove `onMountOpen` and `onMountUnmount` event that was introduced during packaging (originally intended to make analytics work quickly).
* Trigger `recordEnterModal` and `recordCloseModal` event in the `DomainPickerButton` component instead, which is easier to maintain.

Also:
* `onCancel` event has been removed. Clicking **Cancel** button now triggers `onClose` event as well. This is to simplify things.

## Testing instructions

Before testing, run `localStorage.debug = 'calypso.analytics'` to see analytics logs on console.

#### Test `calypso_newsite_modal_open` event
- Opening popover should trigger this event.
- Opening modal should trigger this event.

#### Test `calypso_newsite_modal_close` event

**Fresh Without Selecting Domain**
  1. Open a fresh gutenboarding using `/new?fresh`
  2. Open popover/modal.
  3. Close popover/modal.
  4. This event should be triggered with `selected_domain` containing `undefined`.

**With A Domain Selected**
  1. Open popover/modal.
  2. Select a domain.
  3. Click **Confirm** button.
  4. This event should be triggered with `selected_domain` containing the domain you've just selected.

**With Another Domain Selected**
  1. Re-open popover/modal when you have a domain selected prior to this.
  2. Select another domain.
  3. Click **Confirm** button.
  4. This event should be triggered with the `selected_domain` containing the new domain you've just selected (and not the previous one).

**Test The Different Ways To Close Popover & Modal**
- You can close popover by:
  - Clicking outside
  - Pressing **ESC** key
  - Clicking **Close (X)** button
  - Clicking **Confirm** button

- You can close modal by:
  - Pressing the **ESC** key
  - Clicking **Confirm** button
  - Clicking **Cancel** button

All of the above should trigger the close event.

_Note: Clicking **More Options** button does not fire the close event._.

## Screenshot

![image](https://user-images.githubusercontent.com/1287077/82917424-b1b97c00-9f73-11ea-8914-dcb22584fdfa.png)

Fixes #42649
Related https://github.com/Automattic/wp-calypso/pull/42409 https://github.com/Automattic/wp-calypso/pull/42657
